Add how to register a sidechain page in lisk docs (#1653)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/understand-blockchain/interoperability/sidechain-registration-and-recovery.adoc

A sidechain is terminated for the following reasons:

. The sidechain breaks the <<liveness-condition>>.
. The sidechain is <<sending-malicious-ccms>>.
. The sidechain is <<violating-the-lisk-interoperability-protocol>>.

==== Liveness condition

The liveness condition is defined as follows:

Active sidechains are required to prove their liveness to the mainchain at least once every 30 days.
This is done by including a CCU on the mainchain.

NOTE: *The liveness condition only needs to be fulfilled on the mainchain.*
That means, the mainchain account on the sidechain is not terminated, if the mainchain doesn't send a CCU to the sidechain within 30 days.

Whenever a sidechain has violated the liveness condition, any user can submit a <<terminate-sidechain-for-liveness-command>> on the mainchain, and this will result in the sidechain being terminated.
This rule guarantees that users do not send funds to inactive sidechains and that users who have tokens in a sidechain which stops communicating with the ecosystem can recover their tokens.

==== Sending malicious CCMs

A sidechain is terminated, if it posts a CCU containing one or multiple malicious xref:{url_ccm}[CCMs].

> Certain xref:{url_understand_module}[modules] of a sidechain client have the ability to terminate a sidechain account.

Reviewer: I think it would be helpful to add an example: Token module on the mainchain can terminate a sidechain if it violates escrow, i.e. sends more LSK tokens to the mainchain that were sent to this sidechian.

If a module receives a CCM that it considers to be malicious, it can call the  xref:{url_lip_45_terminatechain}[`terminateChain`] method of the Interoperability module to terminate the sidechain.

> NOTE: Notice that a sidechain terminated for this reason is not terminated in the whole ecosystem.

Reviewer: This is not entirely correct. If the mainchain terminates a sidechain, then this chain is terminated in the whole ecosystem (e.g. when violating LSK escrow). 
However imagine that a sidechain A violates escrow on a sidechain B (by essentially trying to mint native tokens from B), then B will stop communicating with A as you write in the note. The mainchain and other sidechain could still be talking to sidechain A.

For example, it could very well still be active and able to exchange messages with other chains.
However, communication with the sidechain that terminated it is interrupted.

==== Violating the Lisk Interoperability protocol

It is also possible for a sidechain to be terminated, if it violates the Lisk Interoperability protocol.
Note, that if a sidechain uses the Lisk Interoperability module, it is never violate the protocol.

But in case another custom module is used for interoperability in a sidechain, and it sends for example CCUs that don't follow the expected format, this can also lead to the termination of the sidechain.

=== Terminate Sidechain For Liveness command

The {url_typedoc_interop_terminate4liveness}[Terminate Sidechain For Liveness^] command is used to terminate a sidechain that violated the liveness condition.
The timestamp of the last certificate provided by the sidechain is checked against the current timestamp.
If the difference is greater than the maximum allowed timespan, the sidechain is terminated.

This command can only be submitted on the mainchain.

> === Sidechain Terminated CCM

Reviewer: Somewhere in this section I would add that the Sidechain Terminated CCM is sent to a sidechain when it tries to send a message to a terminated sidechain (and of course fails). See here: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0049.md#creation-2.

Otherwise it sounds to me like the mainchain informs all the chains in the ecosystem when a sidechain is terminated.

The role of the sidechain terminated cross-chain message is to inform sidechains that another sidechain has been terminated on the mainchain and is unable to receive messages.
The message contains the ID of the terminated chain as well as the last certified state root of the terminated sidechain (as certified on the mainchain).
This value is used for the creation of the terminated state account (on the sidechain receiving the CCM), allowing state recoveries.

This message allows to inform sidechains about other terminated sidechains efficiently.
Indeed, this message will automatically trigger the creation of the terminated sidechain account as soon as the first message is unable to be delivered.
This also prevents further messages to be sent to already terminated sidechains.

=== Channel Terminated CCM

A "channel terminated" CCM is sent to chains which have been terminated.

The role of the "channel terminated" message is to inform chains that their channel has been terminated on the mainchain.
The chain receiving this message can then also close the channel to the mainchain.

> This is helpful in preventing users from sending transactions to a chain whilst the cross-chain update transaction will be invalid.

Reviewer 2: ```suggestion
This is helpful in preventing users from sending transactions to a chain ,since the cross-chain update transaction will be invalid.
```

A sidechain account can be terminated on a sidechain using the `terminateChain` function exposed by the Interoperability module.
In this case, the state root is generally not available and the terminated state account is created without setting the sidechain state root.
Instead, the account stores the mainchain state root at the time of termination.

A state recovery initialization command can be posted afterwards to set the sidechain state root by giving an inclusion proof against this mainchain state root.
It contains an inclusion proof against the mainchain state root stored in the terminated state account, proving the value of the state root of the terminated sidechain.
Any user on the corresponding sidechain can send a transaction with this command and initiate the state recoveries with respect to the terminated sidechain.
Further information including, the `createTerminatedStateAccount` function can be found in {url_lip_45}[LIP 0045].

NOTE: This command can only be submitted on a sidechain.

> This command has two purposes: It can be used to...

Reviewer 2: ```suggestion
This command has two purposes.
It can be used to perform the following:
```

. ...set the state root of a "terminated state" account on a sidechain, or
. ...to terminate another sidechain directly.

==== Recover State

> The {url_typedoc_interop_state_recovery}[Recover State^] command is used to recover assets (for example fungible and non-fungible tokens) from a terminated sidechain.

Reviewer 2: ```suggestion
The {url_typedoc_interop_state_recovery}[Recover State^] command is used to recover assets (for example, fungible and non-fungible tokens) from a terminated sidechain.
```

[NOTE]
====
The outbox root property is duplicated and additionally stored separately from all other properties of the chain account.
Storing the outbox root with a different substore prefix allows to separate the subtree corresponding to the outbox roots from the rest of the data in the interoperability store.
This choice allows for shorter inclusion proofs for the outbox root, as the other properties of the Interoperability module are not needed to recalculate the state root if the outbox root is known.
In particular, the inclusion proof contained in a CCU from a sidechain posted on the mainchain will contain only one hash.
====

[%collapsible]
====
Key:: Each store key is set to the ID of the partner chain `chainID`.

Value::
Each entry in the "outbox root" substore includes the following property:

* `root`: The root of the underlying Merkle tree of the partner chain outbox.
====

=== Own chain substore

The "own chain" substore stores the name and ID of the respective chain.

On the Lisk *mainchain*, the own chain account is present by default, and set to an object with properties:

[source,js]
----
{
    name: "lisk_mainchain",

>     chainID: 00000000,

Reviewer: Note that `chainID: 00000000` is true only in the mainnet, the testnet and the betanet have different chain IDs. See here: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0037.md#mainchain-chain-id.

I would mention that this object is valid for Lisk mainchain in the mainnet.

* `lastCertificate`: This property holds some information from the certificate posted contained in the last CCU from the partner chain.
It is an object containing the following properties:
** `height`: The height contained in the last certificate from the partner chain.
It is used to validate a certificate{fn_lip53} (certificates must contain block headers with increasing heights).
** `timestamp`: The timestamp contained in the last certificate from the partner chain.
On the mainchain, it is used to check that the sidechain chain fulfills the <<liveness-condition>>.
** `stateRoot`: The state root contained in the last certificate from the partner chain.
It is used to validate the inclusion proof of the cross-chain messages contained in a CCU and to verify the validity of the module recovery command.
** `validatorsHash`: The validators hash contained in the last certificate from the partner chain.
This value always corresponds to the hash calculated from the <<chain-validators-substore>> entry for the partner chain.
* `status`: This property stores the current status of the partner chain account.

> As explained in <<life-cycle-of-a-sidechain>>, there are 3 possible statuses: `active`(0), `registered`(1), and `terminated`(2).

Reviewer: Should be: "there are 3 possible statuses: `registered`(0), `active`(1), and `terminated`(2)."

The default value of this property is `0`, corresponding to the `registered` status.
====

=== Channel substore
The channel substore holds information about the inbox and outbox with other partner chains.

It is initialized on the respective chain as part of the <<sidechain-registration>> process.

* On the mainchain, the store contains channels to all registered sidechains.
* On a sidechain, the store contains a channel to the mainchain.

[%collapsible]
====
Key:: Each store key is set to the ID of the partner chain `chainID`.

Value::
Each entry in the channel substore includes the following properties:

* `inbox`: The data structure containing information about the cross-chain messages received from the partner chain, organized in a regular Merkle tree{fn_lip31_rmt}.
It contains the following properties:
** `root`:  The root of the Merkle tree.
** `appendPath`: An array of hashes necessary to append new data to the tree efficiently.

> ** `size`: The current size of the tree, i.e. the number of cross-chain messages received from the partner chain and processed.

Author: ```suggestion
** `size`: The current size of the tree, i.e. the number of cross-chain messages received from the partner chain that have been processed.
```

* `outbox`: The data structure containing information about the cross-chain messages sent to the partner chain, organized in a regular Merkle tree.
It contains the following properties:
** `root`:  The root of the Merkle tree.
** `appendPath`: An array of hashes necessary to append new data to the tree efficiently.

> ** `size`: The current size of the tree, i.e. the number of cross-chain messages sent to the partner chain.

Author: ```suggestion
** `size`: The current size of the tree, i.e., the number of cross-chain messages sent to the partner chain.
```

For the mainchain account on a sidechain, the `activeValidators` and `certificateThreshold` properties are initialized by the <<register-mainchain-command>>.
For a sidechain account on the mainchain, they are set by the <<register-sidechain-command>>.
====

=== Terminated state substore

A *terminated state* account is created for a terminated chain on the mainchain or another sidechain ...

* ... as part of the `terminateChain` function,
* ... as part of the processing of a <<sidechain-terminated-ccm>>,
* ... as part of the processing of a <<channel-terminated-ccm>>,
* ... or as part of the processing of a <<recover-state>> command.

It is initialized with the information contained in the last sidechain certificate posted on the mainchain.

[%collapsible]
====
Key::
The store key is set to the ID of the terminated chain `chainID`.

Value::
Each entry in the "terminated state" substore includes the following properties:

* `stateRoot`: The state root of the terminated chain.
* `mainchainStateRoot`: The state root of the mainchain at the moment in which the chain was terminated.

> * `initialized`: A boolean value, indicating whether the terminated state account has been initialized, i.e. if the `stateRoot` property has been set.

Author: ```suggestion
* `initialized`: A boolean value, indicating whether the terminated state account has been initialized, i.e., if the `stateRoot` property has been set.
```

= Sidechain Registration & Recovery
Chris Braithwaite <christopher.braithwaite@lightcurve.io>, Mona Bärenfänger <mona@lightcurve.io>
:toc: preamble
:idprefix:
:idseparator: -
:imagesdir: ../../assets/images

> //URLs

Reviewer 2: I think in some places I did this intentionally, because using the ordered list format will break other stylings (`::` for example)

> //URLs

Reviewer 2: I am not aware of places where it was not done intentionally. As far as i am aware this is only done for the recovery steps. But also, for me it doesn't decrease readability on the page.

Additionally, pending messages that were stuck unprocessed in the outbox of the terminated chain can be recovered as well, and their effect is reverted on the sending chain.

NOTE: No connection or communication to the terminated sidechain is required to complete the recovery.

.Recovering tokens and NFTs from terminated sidechains
video::iBVfttn_n-U[youtube,align=center,width=100%,height=400]

[IMPORTANT]
====
For the sidechain recovery process, it is not important who is sending the recovery transactions:
The recovery process happens independently of the sender, and tokens and assets will always be recovered for the correct accounts.

This makes it possible to provide <<recovery-as-a-service>> to users, to further simplify the recovery process.
====

=== Recovery process from the mainchain

On the mainchain, users can perform the following recoveries:

. <<state-recovery, State recovery>>
. <<ccm-recovery, CCM recovery>>

The recovery process for on the mainchain works as follows:

.Recovery process from the mainchain.
image::understand-blockchain/interop/recovery_mainchain.png[align="center",width=550]

[#state-recovery]

> 1.State recovery::

Reviewer 2: For me this breaks the styling with the `::` unfortunately :(
